### PR TITLE
Fix false positives for maps in property-no-unknown

### LINKED ISSUE #5654: Fix false positives for maps in color-no-invalid-hex and property-no-unknown
Less feature known as maps supposed to be supported atm. However both maps init and maps usage produce false positives.

> Which rule, if any, is the bug related to?

`property-no-unknown`
`color-no-invalid-hex`

> What code is needed to reproduce the bug?

```less
#colors() {
  somecolor: red;
  someothercolor: #fff;
}

.myclass {
  color: #colors[somecolor];
}
```

> What Stylelint configuration is needed to reproduce the bug?

reproducable even without it, but if anything..
had to use `postcss-less` following `stylelint` v.14 update logic where there's no more in-built analyzers.

```json
{
  "extends": "stylelint-config-recommended-less",
  "customSyntax": "postcss-less"
}
```

> Which version of Stylelint are you using?

`14.0.0`

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

Stylelint extension @ VSCode, Stylelint as default less linter

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

Yes, it's related to LESS maps.

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors did you get)?

The following warnings were flagged:

```shell
error Unexpected invalid hex color "#colors" (color-no-invalid-hex)
error Unexpected unknown property "somecolor" (property-no-unknown)
error Unexpected unknown property "someothercolor" (property-no-unknown)
```

Related to [this issue](https://github.com/stylelint/stylelint/issues/4834)

### PR DESCRIPTION
**Pull request by strigefleur**
> Which issue, if any, is this issue related to?

Partial fix #5654 

> Is there anything in the PR that needs further explanation?

No, it's self-explanatory.

### CODE REVIEW

**Comment by strigefleur** [comment]
Was going to split into separate PRs but then ended up pushing in the same branch.
Hope that's fine.
In it's current state should fix #5654 whole.

File: lib/utils/isStandardSyntaxDeclaration.js [hunk 1]
```diff
@@ -46,6 +46,17 @@ module.exports = function (decl) {
 		return false;
 	}
 
+	// Less map (e.g. #my-map() { myprop: red; })
+	if (
+		parent &&
+		isRule(parent) &&
+		parent.selector &&
+		parent.selector[0] === '#' &&
+		parent.selector[parent.selector.length - 1] === ')'
```
Comment (ybiquitous) [inline lib/utils/isStandardSyntaxDeclaration.js:55]: [suggestion] Does it make sense to make the check a bit readable as below?

```suggestion
		parent.selector.startsWith('#') &&
		parent.selector.endsWith('()')
```

File: lib/utils/__tests__/isStandardSyntaxHexColor.test.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+'use strict';
+
+const isStandardSyntaxHexColor = require('../isStandardSyntaxHexColor');
+
+describe('isStandardSyntaxHexColor', () => {
+	it('default hex', () => {
+		expect(isStandardSyntaxHexColor('#000000')).toBeTruthy();
+	});
+	it('less map usage', () => {
+		expect(isStandardSyntaxHexColor('#prop[someprop]')).toBeFalsy();
```
Comment (ybiquitous) [inline lib/utils/__tests__/isStandardSyntaxHexColor.test.js:10]: [suggestion] For more accuracy, how about using `toBe()` like this?

```suggestion
		expect(isStandardSyntaxHexColor('#000000')).toBe(true);
	});
	it('less map usage', () => {
		expect(isStandardSyntaxHexColor('#prop[someprop]')).toBe(false);
```

File: lib/utils/isStandardSyntaxHexColor.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
+'use strict';
+
+/**
+ * Check whether a hex color is standard
+ *
+ * @param {string} property
+ * @returns {boolean}
+ */
+module.exports = function (property) {
```
Comment (ybiquitous) [inline lib/utils/isStandardSyntaxHexColor.js:9]: [suggestion] Let's name the function for better debuggability:

```suggestion
module.exports = function isStandardSyntaxHexColor(property) {
```

**Review by ybiquitous — COMMENTED** [review]
@strigefleur Thank you for creating this PR! 😄 

I've suggested some refactoring ideas, but almost LGTM. 👍🏼

**Comment by strigefleur** [comment]
Agreed on every point :)
There's lot of `function (property)` syntax already so probably one day someone will refactor it as well.

**Review by ybiquitous — APPROVED** [review]
@strigefleur Thank you. LGTM! 😄

File: lib/utils/isStandardSyntaxHexColor.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+'use strict';
+
+/**
+ * Check whether a hex color is standard
+ *
+ * @param {string} property
```
Comment (jeddy3) [inline lib/utils/isStandardSyntaxHexColor.js:6]:
```suggestion
 * @param {string} hex
```

File: lib/utils/isStandardSyntaxHexColor.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+ * @returns {boolean}
+ */
+module.exports = function isStandardSyntaxHexColor(property) {
```
Comment (jeddy3) [inline lib/utils/isStandardSyntaxHexColor.js:9]:
```suggestion
module.exports = function isStandardSyntaxHexColor(hex) {
```

File: lib/utils/isStandardSyntaxHexColor.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+	// Less map usage (e.g. .myclass { color: #colors[somecolor]; })
+	if (property.includes('[')) {
```
Comment (jeddy3) [inline lib/utils/isStandardSyntaxHexColor.js:11]:
```suggestion
	if (hex.includes('[')) {
```

File: lib/rules/color-no-invalid-hex/__tests__/index.js [hunk 7]
```diff
@@ -61,6 +61,10 @@ testRule({
 				'font-style: normal;\n' +
 				'}',
 		},
+		{
+			code: 'a { color: #colors[somecolor]; }',
+			description: 'Less map usage',
+		},
```
Comment (jeddy3) [inline lib/rules/color-no-invalid-hex/__tests__/index.js:67]: Let's move this into its own `testRule` case (towards the bottom of the file), which sets to syntax to `postcss-less` as `testRule` without syntax should only include vanilla CSS constructs.

**Review by jeddy3 — CHANGES_REQUESTED** [review]
Thanks for the pull request.

I've requested some changes.

File: lib/rules/color-no-invalid-hex/__tests__/index.js (hunk 7 above)
Comment (strigefleur) [reply lib/rules/color-no-invalid-hex/__tests__/index.js:67]: 👍  Sure, as soon as I have at least some time.

**Review by jeddy3 — APPROVED** [review]
LGTM, thanks!

(I move the test case myself).

**Comment by jeddy3** [comment]
> - Fixed: `property-no-unknown` false positives for maps ([#5690](https://github.com/stylelint/stylelint/pull/5690)).

**Pull request merged**